vertx-mongo-client/src/main/java/io/vertx/ext/mongo/impl/GridFSOutputStreamImpl.java

```diff
@@ -0,0 +1,41 @@
+package io.vertx.ext.mongo.impl;
+
+import com.mongodb.async.SingleResultCallback;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.streams.WriteStream;
+import io.vertx.ext.mongo.GridFSOutputStream;
+
+import java.nio.ByteBuffer;
+
+/**
+ *  Implementation of {@link GridFSOutputStream} which allows streaming with
+ *  Vertx's  {@link io.vertx.core.streams.Pump}
+ *
+ * Adapted from com.github.sth.vertx.mongo.streams.GridFSOutputStreamImpl
+ *
+ * @author <a href="https://github.com/st-h">Steve Hummingbird</a>
+ */
+public class GridFSOutputStreamImpl implements GridFSOutputStream {
+
+    WriteStream<Buffer> writeStream;
+
+    public GridFSOutputStreamImpl(WriteStream<Buffer> writeStream) {
```

Comment:
Committed code for most of your issues except this one. Not quite finished with it. I feel it requires more testing. Red Hat will be keeping me busy until 25 May. I'll take it up again then.